File: CodingDojo/challenges/2022-03-15-longest-common-prefix/solutions/julia/ericwburden/ericwburden.jl
"""
    Base.convert(::Type{Matrix{Char}}, strings::Vector{String})

Converts a `Vector{String}` to a `Matrix{Char}`, where each column contains
the characters from a string in `strings`, truncated to the length of the
shortest string.
"""
function Base.convert(::Type{Matrix{Char}}, strings::Vector{String})
    sorted = sort!(strings, by = length)
    max_length = length(sorted[1])
    return reduce(hcat, collect(s)[1:max_length] for s in strings)
end

"Helper function to identify iterables where all the values are the same"
all_equal(v) = all(x -> x == v[1], v)

"""
    longest_common_prefix(strings::Vector{String})

Given a list of strings, returns a string containing the longest common prefix
among all strings. Returns an empty string if no matching prefix is found.
"""
function longest_common_prefix(strings::Vector{String})
    char_matrix    = convert(Matrix{Char}, strings)
    matching_chars = Iterators.takewhile(all_equal, eachrow(char_matrix))
    return map(x -> first(x), matching_chars) |> join
end


# Tests! -----------------------------------------------------------------------

using Test

@testset "Should return any single letter as-is" begin
    @test longest_common_prefix(["a"]) == "a"
    @test longest_common_prefix(["b"]) == "b"
    @test longest_common_prefix(["z"]) == "z"
end

@testset "Should identify single-letter prefixes" begin
    input = ["apple", "aardvark"]
    @test longest_common_prefix(input) == "a"

    input = ["garbage", "graendal", "gift"]
    @test longest_common_prefix(input) == "g"

    input = ["eagle", "easter", "easy", "egg"]
    @test longest_common_prefix(input) == "e"
end

@testset "Should identify multiple-letter prefixes" begin
    input = ["apple", "apricot"]
    @test longest_common_prefix(input) == "ap"

    input = ["garbage", "garden", "gargantuan"]
    @test longest_common_prefix(input) == "gar"

    input = ["castle", "casting", "castanet", "cast"]
    @test longest_common_prefix(input) == "cast"
end

@testset "Should identify when there is no common prefix" begin
    input = ["apple", "banana"]
    @test longest_common_prefix(input) == ""

    input = ["garbage", "garden", "helicopter"]
    @test longest_common_prefix(input) == ""

    input = ["castle", "antimony", "castanet", "antiquity"]
    @test longest_common_prefix(input) == ""
end

@testset "Should identify repeated instances of a single word" begin
    input = ["apple", "apple"]
    @test longest_common_prefix(input) == "apple"

    input = ["apple", "apple", "apple"]
    @test longest_common_prefix(input) == "apple"
end

@testset "Should work on large inputs" begin
    input = fill("variadic", 10000)
    @test longest_common_prefix(input) == "variadic"

    append!(input, fill("variable", 10000))
    append!(input, fill("variety", 10000))
    @test longest_common_prefix(input) == "vari"

    push!(input, "a")
    @test longest_common_prefix(input) == ""
end

File: CodingDojo/challenges/2022-03-15-longest-common-prefix/solutions/rust/ericwburden/ericwburden.rs
fn common_prefix(left: String, right: &str) -> String {
    let mut prefix = String::from("");
    for (l, r) in left.chars().zip(right.chars()) {
        if l == r { prefix.push(l); } else { break; }
    }
    prefix
}

fn longest_common_prefix(strings: &[&str]) -> String {
    strings
        .iter()
        .copied()
        .fold(strings[0].to_string(), common_prefix)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn testing() {
        let x = common_prefix("g".to_string(), "gargantuan");
        println!("{}", x);
    }

    #[test]
    fn single_letter() {
        //! Should return any single letter as-is
        assert_eq!(longest_common_prefix(&["a"]), "a");
        assert_eq!(longest_common_prefix(&["g"]), "g");
        assert_eq!(longest_common_prefix(&["z"]), "z");
        assert_eq!(longest_common_prefix(&["T"]), "T");
    }

    #[test]
    fn single_letter_prefix() {
        //! Should identify single-letter prefixes
        let input = ["apple", "aardvark"];
        assert_eq!(longest_common_prefix(&input), "a");

        let input = ["garbage", "graendal", "gift"];
        assert_eq!(longest_common_prefix(&input), "g");

        let input = ["eagle", "easter", "easy", "egg"];
        assert_eq!(longest_common_prefix(&input), "e");
    }

    #[test]
    fn multiple_letter_prefix() {
        //! Should identify multiple-letter prefixes
        let input = ["apple", "apricot"];
        assert_eq!(longest_common_prefix(&input), "ap");

        let input = ["garbage", "garden", "gargantuan"];
        assert_eq!(longest_common_prefix(&input), "gar");

        let input = ["castle", "casting", "castanet", "cast"];
        assert_eq!(longest_common_prefix(&input), "cast");
    }

    #[test]
    fn no_letter_prefix() {
        //! Should identify when there is no common prefix
        let input = ["apple", "banana"];
        assert_eq!(longest_common_prefix(&input), "");

        let input = ["garbage", "garden", "helicopter"];
        assert_eq!(longest_common_prefix(&input), "");

        let input = ["castle", "antimony", "castanet", "antiquity"];
        assert_eq!(longest_common_prefix(&input), "");
    }

    #[test]
    fn repeated_words() {
        //! Should identify when there are repeated instances of a single word
        let input = ["apple", "apple"];
        assert_eq!(longest_common_prefix(&input), "apple");

        let input = ["apple", "apple", "apple"];
        assert_eq!(longest_common_prefix(&input), "apple");
    }

    #[test]
    fn large_inputs() {
        //! Should work on large inputs
        let mut input = vec!["variadic"; 10_000];
        assert_eq!(longest_common_prefix(&input), "variadic");

        input.extend(&["variable"; 1000]);
        input.extend(&["variety"; 1000]);
        assert_eq!(longest_common_prefix(&input), "vari");

        input.push("a");
        assert_eq!(longest_common_prefix(&input), "");
    }
}
